src/stores/Frame/FrameStore.ts

```diff
@@ -2821,4 +2837,85 @@ export class FrameStore {
         this.fittingResultRegionParams = [];
         this.fittingLog = "";
     };
+
+    @action setIsPreview = (isPreview: boolean) => {
+        this.isPreview = isPreview;
+    };
+
+    @action setPreviewPVRasterData = (previewPVRasterData: Float32Array, skipUpdatePreviewData: boolean = false) => {
+        this.previewPVRasterData = previewPVRasterData;
+        // if skipUpdatePreviewData is false, the code after the yield keyword in the updatePreviewData function will be executed by calling the next() function.
+        !skipUpdatePreviewData && this.updatePreviewDataGenerator.next();
+    };
+
+    public updatePreviewDataGenerator: Generator;
+
+    // The incoming data will be decompressed using zfp WebWorker instance, which runs in a different thread. WebWorker instance has an onmessage event listener function. Generator function is used here to wait for the decompression of rasterData to complete.
+    public *updatePreviewData(previewData: CARTA.PvPreviewData) {
+        // Old values before updating to the new frameInfo
+        const oldAspectRatio = this.aspectRatio;
+        const oldHeight = this.frameInfo.fileInfoExtended.height;
+        const oldWidth = this.frameInfo.fileInfoExtended.width;
+
+        const compressedArray = previewData.imageData;
+        const nanEncodings32 = new Int32Array(previewData.nanEncodings.slice(0).buffer);
+        let compressedView = new Uint8Array(Math.max(compressedArray.byteLength, previewData.width * previewData.height * 4));
+        compressedView.set(compressedArray);
+
+        const eventArgs = {
+            fileId: PREVIEW_PV_FILEID,
+            channel: 0,
+            stokes: 0,
+            width: previewData.width,
+            subsetHeight: previewData.height,
+            subsetLength: compressedArray.byteLength,
+            compression: previewData.compressionQuality,
+            nanEncodings: nanEncodings32,
+            tileCoordinate: 0,
+            layer: 0,
+            requestId: 0,
+            previewId: previewData.previewId
+        };
+
+        // Using the 'yield' keyword of generator functions to wait for decompressed raster data from other WebWorker thread.
+        // next() will be called in setPreviewPVRasterData, which will be called in the onmessage() function after receiving the decompressed data from other worker thread.
+        yield TileService.Instance.decompressPreviewRasterData(compressedView, eventArgs, previewData, nanEncodings32);
```

Comment:
I suggest we adjust the function to `decompressPreviewRasterData(fileId: number, previewData: CARTA.PvPreviewData)` and move lines 2860-2878 into `TileService`.